### PR TITLE
Added test for recursive delete scenario for F-71A

### PR DESCRIPTION
**Pull request by aditipande96**
Signed-off-by: Aditi Pande <aditi.pande@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/csm/rest/test_capacity_quota.py [hunk 1]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+        Verify user capacity after recursively deleting all objects in bucket
+        """
+        test_case_name = cortxlogging.get_frame()
+        test_cfg = self.csm_conf["test_44411"]
+        max_size = test_cfg["max_size"]
+        random_size = self.csm_obj.random_gen.randrange(1, max_size)
+        num_objects = test_cfg["num_objects"]
+        random_objects = self.csm_obj.random_gen.randrange(1, num_objects)
+        self.log.info("Step 1:Create and upload objects in a bucket")
+        for _ in range(0, random_objects):
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2708]: Can u log to print which Iteration is going on?

File: tests/csm/rest/test_capacity_quota.py [hunk 2]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+            obj_name = f'{self.obj_name_prefix}{time.perf_counter_ns()}'
+            resp = s3_misc.create_put_objects(obj_name, self.bucket,
+                                          self.akey, self.skey,
+                                 object_size=int(random_size/1024), block_size="1K")
+        assert resp, "Put Object failed"
+        import pdb;pdb.set_trace()
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2714]: Can u remove this?

File: tests/csm/rest/test_capacity_quota.py [hunk 3]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+        """
+        test_case_name = cortxlogging.get_frame()
+        test_cfg = self.csm_conf["test_44411"]
+        max_size = test_cfg["max_size"]
+        random_size = self.csm_obj.random_gen.randrange(1, max_size)
+        num_objects = test_cfg["num_objects"]
+        random_objects = self.csm_obj.random_gen.randrange(1, num_objects)
+        self.log.info("Step 1:Create and upload objects in a bucket")
+        for _ in range(0, random_objects):
+            obj_name = f'{self.obj_name_prefix}{time.perf_counter_ns()}'
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2709]: Add log specify object name , bucket , akey, skey and size

File: tests/csm/rest/test_capacity_quota.py [hunk 4]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+            resp = s3_misc.create_put_objects(obj_name, self.bucket,
+                                          self.akey, self.skey,
+                                 object_size=int(random_size/1024), block_size="1K")
+        assert resp, "Put Object failed"
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2713]: Can u change assertion error log to f"Put object {obj_name} failed"

File: tests/csm/rest/test_capacity_quota.py [hunk 5]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+        import pdb;pdb.set_trace()
+        self.log.info("Step 2: Delete all uploaded objects recursively")
+        assert s3_misc.delete_objects(self.bucket, self.akey, self.skey), "Delete object Failed"
+        self.log.info("Step 3: Get user capacity stats")
+        res, resp = self.csm_obj.verify_user_capacity(self.user_id, 0,
+                                0, 0, aligned=True)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2719]: Can u log resp also?

File: tests/csm/rest/test_capacity_quota.py [hunk 6]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+        self.log.info("Step 1:Create and upload objects in a bucket")
+        for _ in range(0, random_objects):
+            obj_name = f'{self.obj_name_prefix}{time.perf_counter_ns()}'
+            resp = s3_misc.create_put_objects(obj_name, self.bucket,
+                                          self.akey, self.skey,
+                                 object_size=int(random_size/1024), block_size="1K")
+        assert resp, "Put Object failed"
+        import pdb;pdb.set_trace()
+        self.log.info("Step 2: Delete all uploaded objects recursively")
+        assert s3_misc.delete_objects(self.bucket, self.akey, self.skey), "Delete object Failed"
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2716]: Can u add NOTE here as we will need to revisit this test.

File: tests/csm/rest/test_capacity_quota.py [hunk 7]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+    @pytest.mark.tags('TEST-44411')
+    def test_44411(self):
+        """
+        Verify user capacity after recursively deleting all objects in bucket
+        """
+        test_case_name = cortxlogging.get_frame()
+        test_cfg = self.csm_conf["test_44411"]
+        max_size = test_cfg["max_size"]
+        random_size = self.csm_obj.random_gen.randrange(1, max_size)
+        num_objects = test_cfg["num_objects"]
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2705]: This is max_objects as we are selecting num_objects randomly.

File: tests/csm/rest/test_capacity_quota.py [hunk 8]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+        random_objects = self.csm_obj.random_gen.randrange(1, num_objects)
+        self.log.info("Step 1:Create and upload objects in a bucket")
+        for _ in range(0, random_objects):
+            obj_name = f'{self.obj_name_prefix}{time.perf_counter_ns()}'
+            resp = s3_misc.create_put_objects(obj_name, self.bucket,
+                                          self.akey, self.skey,
+                                 object_size=int(random_size/1024), block_size="1K")
+        assert resp, "Put Object failed"
+        import pdb;pdb.set_trace()
+        self.log.info("Step 2: Delete all uploaded objects recursively")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2715]: Get capacity before this step to check written objects are reflected correctly.

File: tests/csm/rest/test_capacity_quota.py [hunk 9]
```diff
@@ -2688,3 +2688,42 @@ def test_43862(self):
...
+        test_cfg = self.csm_conf["test_44411"]
+        max_size = test_cfg["max_size"]
+        random_size = self.csm_obj.random_gen.randrange(1, max_size) #specified in bytes
+        num_objects = test_cfg["max_objects"]
+        random_objects = self.csm_obj.random_gen.randrange(1, num_objects)
+        self.log.info("Step 1: Create and upload objects in a bucket")
+        for num in range(0, random_objects):
+            self.log.info("Creating object : %s", num)
+            obj_name = f'{self.obj_name_prefix}{time.perf_counter_ns()}'
+            self.log.info("Creating object %s of size %s in bucket %s ", obj_name,
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2711]: Mention size in logs

File: tests/csm/rest/test_capacity_quota.py [hunk 10]
```diff
@@ -2688,3 +2688,42 @@ def test_43862(self):
...
+                                  int(random_size/1024), self.bucket)
+            resp = s3_misc.create_put_objects(obj_name, self.bucket,
+                                          self.akey, self.skey,
+                                 object_size=int(random_size/1024), block_size="1K")
+        assert resp, f"Put object {obj_name} failed"
+        self.log.info("Step 1: Perform & Verify GET API to get capacity usage stats")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2717]: Step number

File: tests/csm/rest/test_capacity_quota.py [hunk 11]
```diff
@@ -2688,3 +2688,42 @@ def test_43862(self):
...
+        resp = self.csm_obj.get_user_capacity_usage("user", self.user_id)
+        assert resp.status_code == HTTPStatus.OK, \
+                "Status code check failed for get capacity"
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:2720]: Can u validate the response?

File: tests/csm/rest/test_capacity_quota.py (hunk 1 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py [hunk 12]
```diff
@@ -2688,3 +2688,34 @@ def test_43862(self):
...
+    @pytest.mark.parallel
+    @pytest.mark.tags('TEST-44411')
+    def test_44411(self):
+        """
+        Verify user capacity after recursively deleting all objects in bucket
+        """
+        test_case_name = cortxlogging.get_frame()
+        test_cfg = self.csm_conf["test_44411"]
+        max_size = test_cfg["max_size"]
+        random_size = self.csm_obj.random_gen.randrange(1, max_size)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 2 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 3 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:2710]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 4 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 6 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 9 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: added

File: tests/csm/rest/test_capacity_quota.py (hunk 11 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: added validation

**Review by kachhwahadivya — APPROVED** [review]

**Comment by aditipande96** [comment]
PASSED tests/csm/rest/test_capacity_quota.py::TestCapacityQuota::test_44411

========= 2339 tests collected in 10.50s ===========

**Review by ketanarlulkar — APPROVED** [review]

**Pull request merged**